File: apps/qt-ai-cli/src/commands/check.ts
import { existsSync, readFileSync, writeFileSync } from "node:fs";
import { resolve } from "node:path";
import {
  aiParagraphsOf, labeledAiRepairPayload, parseLabeledAiTranslation,
} from "@/lib/ai-paragraphs";
import {
  checkAiTranslationViolations,
  type TranslationViolation,
} from "@/lib/ai-translation";
import {
  loadState, loadStoryConfig, readRawChapter, saveState, storyPaths, workFile,
} from "../story-fs.ts";

export interface CheckResult {
  pass: boolean;
  missing: number[];
  violations: TranslationViolation[];
  ratio: number;
  /**
   * Hết vòng review mà chỉ còn vi phạm rule (đủ đoạn, đủ dài): giống web,
   * chương được coi là pass kèm cảnh báo thay vì chết ở error.
   */
  acceptedWithWarnings: boolean;
  /** Mọi vấn đề còn lại, đã quy về nhãn [[n]] cho người đọc — accept ghi vào state.warnings. */
  issues: string[];
  escalatedToError: boolean;
  reviewPath?: string;
}

/**
 * `violations` chứa `line` theo văn bản đã lắp (finalText): mỗi đoạn còn lại
 * (paragraph không undefined trong `parsed`) chiếm 1 dòng, cách nhau 1 dòng
 * trống — đoạn thứ k (tính trên các đoạn CÒN LẠI, 1-based) nằm ở dòng 2k−1.
 * Muốn agent sửa đúng chỗ trong draft, phải quy dòng đó về nhãn [[n]] gốc:
 * lấy vị trí k rồi tra ngược `definedIndices` (chỉ số 0-based trong mảng
 * `parsed` gốc, đã bỏ các đoạn thiếu) để ra nhãn thật.
 */
function violationLabel(parsed: Array<string | undefined>, item: TranslationViolation): number {
  const definedIndices = parsed
    .map((paragraph, index) => (paragraph !== undefined ? index : -1))
    .filter((index) => index >= 0);
  const position = Math.ceil(item.line / 2); // vị trí đoạn trong finalText (1-based)
  const originalIndex = definedIndices[position - 1];
  return originalIndex !== undefined ? originalIndex + 1 : position;
}

function buildViolationsSection(
  id: string,
  parsed: Array<string | undefined>,
  violations: TranslationViolation[],
): string {
  const list = violations
    .map((item) => `- [[${violationLabel(parsed, item)}]] (dòng ${item.line} trong bản lắp): ${item.message} — "${item.text}"`)
    .join("\n");
  return [
    `# Vi phạm rule — sửa tối thiểu ngay trong work/${id}.draft.md`,
    "",
    "Sửa ĐÚNG TẠI CHỖ trong work/" + id + ".draft.md: với mỗi vi phạm dưới đây, tìm đoạn có nhãn [[n]] tương ứng, " +
      "chỉ đổi đúng từ/cụm gây lỗi, GIỮ NGUYÊN toàn bộ nhãn [[n]] hiện có và không viết lại hay chau chuốt các đoạn khác.",
    "",
    list,
  ].join("\n");
}

export function assembleDraft(root: string, id: string): {
  paragraphs: string[];
  parsed: Array<string | undefined>;
  finalText: string;
} {
  const paths = storyPaths(resolve(root));
  const draftPath = workFile(paths, id, "draft");
  if (!existsSync(draftPath)) {
    throw new Error(`Chưa có bản dịch ${draftPath} — agent phải ghi draft trước khi check.`);
  }
  const paragraphs = aiParagraphsOf(readRawChapter(paths, id));
  const parsed =
    parseLabeledAiTranslation(readFileSync(draftPath, "utf8"), paragraphs.length) ??
    Array.from({ length: paragraphs.length }, () => undefined);
  const finalText = parsed
    .filter((paragraph): paragraph is string => paragraph !== undefined)
    .join("\n\n");
  return { paragraphs, parsed, finalText };
}

export function runCheck(root: string, id: string): CheckResult {
  const paths = storyPaths(resolve(root));
  const state = loadState(paths);
  const chapter = state.chapters[id];
  if (!chapter) throw new Error(`Không có chương ${id} trong state.json.`);
  const story = loadStoryConfig(paths);
  const { paragraphs, parsed, finalText } = assembleDraft(root, id);

  const missing = parsed
    .map((paragraph, index) => (paragraph === undefined ? index + 1 : 0))
    .filter((label) => label > 0);
  const violations = checkAiTranslationViolations(finalText, story.checkRules);
  const rawLength = paragraphs.join("").replace(/\s/g, "").length;
  const translatedLength = finalText.replace(/\s/g, "").length;
  const ratio = rawLength > 0 ? translatedLength / rawLength : 1;
  const tooShort = ratio < state.settings.minLengthRatio;
  const clean = missing.length === 0 && violations.length === 0 && !tooShort;

  const issues: string[] = [
    ...missing.map((label) => `[[${label}]] thiếu đoạn`),
    ...violations.map((item) => `[[${violationLabel(parsed, item)}]] ${item.message} — "${item.text}"`),
    ...(tooShort ? [`Quá ngắn: tỉ lệ ký tự dịch/raw ${ratio.toFixed(2)} < ${state.settings.minLengthRatio}`] : []),
  ];

  let pass = clean;
  let acceptedWithWarnings = false;
  let escalatedToError = false;
  let reviewPath: string | undefined;

  if (!clean) {
    if (chapter.reviewRound >= state.settings.maxReviewRounds) {
      if (missing.length === 0 && !tooShort) {
        // Giống web: hết vòng soát mà chỉ còn vi phạm rule thì vẫn chốt, kèm
        // cảnh báo cho người dùng xem sau. Error chỉ dành cho thiếu đoạn/quá ngắn.
        pass = true;
        acceptedWithWarnings = true;
      } else {
        escalatedToError = true;
        state.chapters[id] = {
          ...chapter,
          status: "error",
          reason: `Quá ${state.settings.maxReviewRounds} vòng review vẫn chưa đạt (thiếu ${missing.length} đoạn, ${violations.length} vi phạm, ratio ${ratio.toFixed(2)}).`,
          updatedAt: Date.now(),
        };
        saveState(paths, state);
      }
    } else {
      const sections: string[] = [];
      if (missing.length > 0) {
        sections.push(
          `# Đoạn còn thiếu — dịch bổ sung các đoạn dưới đây rồi CHÈN vào work/${id}.draft.md, ` +
            "giữ đúng nhãn [[n]] cho từng đoạn, không sửa các đoạn đã có\n\n" +
            labeledAiRepairPayload(paragraphs, missing.map((label) => label - 1)),
        );
      }
      if (violations.length > 0) {
        sections.push(buildViolationsSection(id, parsed, violations));
      }
      if (tooShort) {
        sections.push(
          `# Bản dịch quá ngắn\n\nTỉ lệ ký tự dịch/raw = ${ratio.toFixed(2)} < ${state.settings.minLengthRatio}. Rà từng đoạn xem có bị tóm tắt/lược ý; dịch đủ 100% nội dung.`,
        );
      }
      reviewPath = workFile(paths, id, "review");
      writeFileSync(reviewPath, `${sections.join("\n\n---\n\n")}\n`, "utf8");
      state.chapters[id] = { ...chapter, reviewRound: chapter.reviewRound + 1, updatedAt: Date.now() };
      saveState(paths, state);
    }
  }

  writeFileSync(
    workFile(paths, id, "check"),
    `${JSON.stringify(
      {
        pass, acceptedWithWarnings, missing, violationCount: violations.length, ratio,
        reviewRound: (state.chapters[id] ?? chapter).reviewRound,
        checkedAt: Date.now(),
      },
      null, 2,
    )}\n`,
    "utf8",
  );
  return {
    pass, missing, violations, ratio, acceptedWithWarnings, issues, escalatedToError,
    ...(reviewPath ? { reviewPath } : {}),
  };
}

File: crates/qt-ai-core/src/commands/check.rs
use crate::check::{check_violations, Violation};
use crate::error::{CoreError, Result};
use crate::paragraphs::{labeled_repair_payload, paragraphs_of, parse_labeled_translation};
use crate::story_fs::{
    load_state, load_story_config, now_ms, read_raw_chapter, read_text, resolve_root, save_state, story_paths,
    work_file, write_text, ChapterState, ChapterStatus, WorkKind,
};
use std::path::{Path, PathBuf};

#[derive(Debug)]
pub struct CheckResult {
    pub pass: bool,
    /// Nhãn 1-based của đoạn thiếu.
    pub missing: Vec<usize>,
    pub violations: Vec<Violation>,
    pub ratio: f64,
    /// Hết vòng review mà chỉ còn vi phạm rule (đủ đoạn, đủ dài): giống web, chương pass kèm cảnh báo.
    pub accepted_with_warnings: bool,
    /// Mọi vấn đề còn lại, quy về nhãn [[n]] — accept ghi vào state.warnings.
    pub issues: Vec<String>,
    pub escalated_to_error: bool,
    pub review_path: Option<PathBuf>,
}

pub struct Draft {
    pub paragraphs: Vec<String>,
    pub parsed: Vec<Option<String>>,
    pub final_text: String,
}

/// `violations[i].line` tính trên final_text (các đoạn CÒN LẠI, mỗi đoạn 1 dòng cách 1 dòng trống:
/// đoạn thứ k nằm ở dòng 2k−1). Quy về nhãn [[n]] gốc qua danh sách chỉ số đoạn không thiếu.
fn violation_label(parsed: &[Option<String>], violation: &Violation) -> usize {
    let defined: Vec<usize> = parsed.iter().enumerate().filter(|(_, p)| p.is_some()).map(|(i, _)| i).collect();
    let position = (violation.line + 1) / 2; // Math.ceil(line / 2)
    defined.get(position.saturating_sub(1)).map(|index| index + 1).unwrap_or(position)
}

fn build_violations_section(id: &str, parsed: &[Option<String>], violations: &[Violation]) -> String {
    let list = violations
        .iter()
        .map(|item| {
            format!(
                "- [[{}]] (dòng {} trong bản lắp): {} — \"{}\"",
                violation_label(parsed, item),
                item.line,
                item.message,
                item.text
            )
        })
        .collect::<Vec<_>>()
        .join("\n");
    [
        format!("# Vi phạm rule — sửa tối thiểu ngay trong work/{id}.draft.md"),
        String::new(),
        format!("Sửa ĐÚNG TẠI CHỖ trong work/{id}.draft.md: với mỗi vi phạm dưới đây, tìm đoạn có nhãn [[n]] tương ứng, chỉ đổi đúng từ/cụm gây lỗi, GIỮ NGUYÊN toàn bộ nhãn [[n]] hiện có và không viết lại hay chau chuốt các đoạn khác."),
        String::new(),
        list,
    ]
    .join("\n")
}

pub fn assemble_draft(root: &Path, id: &str) -> Result<Draft> {
    let paths = story_paths(&resolve_root(root));
    let draft_path = work_file(&paths, id, WorkKind::Draft);
    if !draft_path.exists() {
        return Err(CoreError::InvalidState(format!(
            "Chưa có bản dịch {} — agent phải ghi draft trước khi check.",
            draft_path.display()
        )));
    }
    let paragraphs = paragraphs_of(&read_raw_chapter(&paths, id)?);
    let parsed = parse_labeled_translation(&read_text(&draft_path)?, paragraphs.len())
        .unwrap_or_else(|| vec![None; paragraphs.len()]);
    let final_text = parsed.iter().flatten().cloned().collect::<Vec<_>>().join("\n\n");
    Ok(Draft { paragraphs, parsed, final_text })
}

fn char_count_no_ws(text: &str) -> usize {
    text.chars().filter(|c| !c.is_whitespace()).count()
}

pub fn run_check(root: &Path, id: &str) -> Result<CheckResult> {
    let paths = story_paths(&resolve_root(root));
    let mut state = load_state(&paths)?;
    let chapter = state
        .chapters
        .get(id)
        .cloned()
        .ok_or_else(|| CoreError::StoryNotFound(format!("Không có chương {id} trong state.json.")))?;
    let story = load_story_config(&paths)?;
    let Draft { paragraphs, parsed, final_text } = assemble_draft(root, id)?;

    let missing: Vec<usize> =
        parsed.iter().enumerate().filter(|(_, p)| p.is_none()).map(|(i, _)| i + 1).collect();
    let violations = check_violations(&final_text, &story.check_rules, story.genre.setting);
    let raw_length = char_count_no_ws(&paragraphs.concat());
    let translated_length = char_count_no_ws(&final_text);
    let ratio = if raw_length > 0 { translated_length as f64 / raw_length as f64 } else { 1.0 };
    let too_short = ratio < state.settings.min_length_ratio;
    let clean = missing.is_empty() && violations.is_empty() && !too_short;

    let mut issues: Vec<String> = missing.iter().map(|label| format!("[[{label}]] thiếu đoạn")).collect();
    issues.extend(
        violations
            .iter()
            .map(|v| format!("[[{}]] {} — \"{}\"", violation_label(&parsed, v), v.message, v.text)),
    );
    if too_short {
        issues.push(format!(
            "Quá ngắn: tỉ lệ ký tự dịch/raw {ratio:.2} < {}",
            state.settings.min_length_ratio
        ));
    }

    let mut pass = clean;
    let mut accepted_with_warnings = false;
    let mut escalated_to_error = false;
    let mut review_path = None;

    if !clean {
        if chapter.review_round >= state.settings.max_review_rounds {
            if missing.is_empty() && !too_short {
                // Giống web: hết vòng soát mà chỉ còn vi phạm rule thì vẫn chốt, kèm cảnh báo.
                pass = true;
                accepted_with_warnings = true;
            } else {
                escalated_to_error = true;
                state.chapters.insert(
                    id.to_string(),
                    ChapterState {
                        status: ChapterStatus::Error,
                        reason: Some(format!(
                            "Quá {} vòng review vẫn chưa đạt (thiếu {} đoạn, {} vi phạm, ratio {ratio:.2}).",
                            state.settings.max_review_rounds,
                            missing.len(),
                            violations.len()
                        )),
                        updated_at: now_ms(),
                        ..chapter.clone()
                    },
                );
                save_state(&paths, &state)?;
            }
        } else {
            let mut sections: Vec<String> = Vec::new();
            if !missing.is_empty() {
                let zero_based: Vec<usize> = missing.iter().map(|label| label - 1).collect();
                sections.push(format!(
                    "# Đoạn còn thiếu — dịch bổ sung các đoạn dưới đây rồi CHÈN vào work/{id}.draft.md, giữ đúng nhãn [[n]] cho từng đoạn, không sửa các đoạn đã có\n\n{}",
                    labeled_repair_payload(&paragraphs, &zero_based)
                ));
            }
            if !violations.is_empty() {
                sections.push(build_violations_section(id, &parsed, &violations));
            }
            if too_short {
                sections.push(format!(
                    "# Bản dịch quá ngắn\n\nTỉ lệ ký tự dịch/raw = {ratio:.2} < {}. Rà từng đoạn xem có bị tóm tắt/lược ý; dịch đủ 100% nội dung.",
                    state.settings.min_length_ratio
                ));
            }
            let path = work_file(&paths, id, WorkKind::Review);
            write_text(&path, &format!("{}\n", sections.join("\n\n---\n\n")))?;
            review_path = Some(path);
            state.chapters.insert(
                id.to_string(),
                ChapterState { review_round: chapter.review_round + 1, updated_at: now_ms(), ..chapter.clone() },
            );
            save_state(&paths, &state)?;
        }
    }

    let review_round = state.chapters.get(id).map(|c| c.review_round).unwrap_or(chapter.review_round);
    let report = serde_json::json!({
        "pass": pass, "acceptedWithWarnings": accepted_with_warnings, "missing": missing,
        "violationCount": violations.len(), "ratio": ratio, "reviewRound": review_round, "checkedAt": now_ms(),
    });
    write_text(
        &work_file(&paths, id, WorkKind::Check),
        &format!("{}\n", serde_json::to_string_pretty(&report).unwrap()),
    )?;

    Ok(CheckResult { pass, missing, violations, ratio, accepted_with_warnings, issues, escalated_to_error, review_path })
}
